File: backend/app/routers/study.py
from typing import Literal, Optional

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from ..dependencies import get_db
from ..services.study_summary import build_study_scope_summary


router = APIRouter()


@router.get("/study/summary")
def get_study_summary(
    scope_type: Literal["group", "tag", "collection", "pack"],
    group_id: Optional[int] = None,
    collection_id: Optional[int] = None,
    tag: Optional[str] = None,
    pack_guid: Optional[str] = None,
    db: Session = Depends(get_db)
):
    return build_study_scope_summary(
        db,
        scope_type,
        group_id=group_id,
        collection_id=collection_id,
        tag=tag,
        pack_guid=pack_guid
    )

File: frontend/src/api/study.js
import { requestJson } from "./http";


function firstDefined(...values) {
  return values.find(value => value !== undefined && value !== null);
}


export function getStudySummary(scope = {}) {
  const params = new URLSearchParams();
  const scopeType = scope.scopeType || scope.type;

  if (scopeType) {
    params.set("scope_type", scopeType);
  }

  if (scopeType === "group") {
    const groupId = firstDefined(scope.groupId, scope.group_id, scope.id);
    if (groupId !== undefined) params.set("group_id", String(groupId));
  }

  if (scopeType === "collection") {
    const collectionId = firstDefined(
      scope.collectionId,
      scope.collection_id,
      scope.id
    );
    if (collectionId !== undefined) {
      params.set("collection_id", String(collectionId));
    }
  }

  if (scopeType === "tag") {
    const tag = firstDefined(scope.tag, scope.key, scope.id, scope.label);
    if (tag) params.set("tag", String(tag));
  }

  if (scopeType === "pack") {
    const packGuid = firstDefined(scope.packGuid, scope.pack_guid, scope.id);
    if (packGuid) params.set("pack_guid", String(packGuid));
  }

  const query = params.toString();

  return requestJson(`/study/summary${query ? `?${query}` : ""}`);
}
